### DISCUSSION TITLE
一种在抢票时可能会节省过验证码时间的方法

### DISCUSSION THREAD

**Discussion by gpc123456** [mikumifa/biliTickerBuy#351] [聊天专区]
在正式抢票前，先使用脚本抢一张已经开票过但售罄的票，随后再使用定时抢票来抢这次要抢的票，有比较大的概率会进入请求的白名单，直接跳过验证码过程，可以节约大量的过验证码的时间。大家有机会的话可以试一下，分享一下我进行测试的方案：
- 先抢BW+BML联票(￥980，已售罄)，log显示需要过验证码，通过验证码后再发一个请求后退出
- 随后使用另一票种：BW+BML联票(￥1308)进行测试，发现抢票时无需再过验证码，能够直接拿到token

这期间只要不更换IP，这个状态能持续大概10分钟的时间

**Comment by Forestchenglin** [comment]
这样会被B站封吗？

**Reply by gpc123456** [reply]
目前来看只要不是向未开票的票种提前发请求，而是向已开票售罄的票种发请求的话是不会封的。

可以在第一个tab配置票种时，选择“已售罄”的票，而不是选“不可售”的票

**Comment by mikumifa** [comment]
天才

**Comment by mikumifa** [comment]
感觉直接继承到代码上就行了

**Reply by GoldenglowSusie** [reply]
是不是可以在定时抢票开始的前一段时间自动执行这个操作（

**Reply by mikumifa** [reply]
是这样的

**Comment by GoldenglowSusie** [comment]
简直是天才，期待大佬pr🥰

**Comment by gpc123456** [comment]
仓库作者觉得可行的话，我就先开个Issue，明天把这块代码完善一下

**Comment by Cloud1069** [comment]
可是这种只能在开票后使用吧，正常B站只会在即将开售的里面上验证码，已经发售完毕的是不会有验证码的吧

**Reply by gpc123456** [reply]
已发售完热门票种的拿token也会要求过验证码的，您可以试一下￥980的BW+BML联票，使用手动过码或者查看log，发现也会需要过验证码。

如果您这段时间有通过验证码的记录，您的IP可能在白名单中，这样就可以直接开始抢票

**Reply by Cloud1069** [reply]
我能想到的是，是否可以多账号在同一IP使用这同一个白名单，就是在发售时只需要解一次码，后面其他的task就不需要再次解码

**Reply by gpc123456** [reply]
这个不太确定，您可以在您本地连接固定的网络，通过切换脚本中的账号来发送抢票请求试一下。

由于我目前只有一个账号，所以还测试不了，抱歉啦~

**Reply by Cloud1069** [reply]
我刚才试了一下联票确实请求了验证码后进入白名单，一般这种checkpoint都是按时效或者请求次数算的。但是bw页面等到二开那天会改变吧，我不确定到时候是否可以使用同一个方法，是否可以用请求bml来通过验证码去bypass?

**Reply by Cloud1069** [reply]
我发现任务开始后又停不下来了

**Reply by SouthSeafireworks** [reply]
可以的啊，就是到时候bml要是也刷新到新票池的话，那就没法验证了，但是第二波bml肯定是可以可以用第二波bw来验证

**Reply by gpc123456** [reply]
我试了，这样不行，可能每个票务id有独立的白名单，要是改变了的话就只能按部就班过验证码了

**Reply by Cloud1069** [reply]
如果bw/bml的验证码不可以混用的话，就只能尝试是否可以验证一次bypass掉不同账号下的验证码了

**Comment by Colin-Chen630** [comment]
我也遇到了。我第一批开票时，脚本+手抢。我后来查看日志发现脚本在第一次请求触发验证后，后面请求一直429。但是我手机直接没有经过验证码直接就可以下单了。我当时电脑连的是手机热点，这样两台设备是不是ip一致，正好符合楼主说的情况。不太了解这方面知识，有无大佬解答。

**Reply by gpc123456** [reply]
脚本429应该是由于抢票的人太多了，没有挤进去。

一般app进提交订单页面都是要过一个验证码的，如果如您所说没有出现验证码，我猜测应该是脚本通过验证码后，ip暂时进入了白名单的原因。

电脑连接手机热点的话，两台设备对外都是手机流量的ip，这与上面的现象也相符。感谢您提供的反馈！

**Reply by Colin-Chen630** [reply]
我昨天跟朋友讨论，就想到过这个验证码白名单的事，感谢大佬帮忙确认这个想法

**Reply by SouthSeafireworks** [reply]
429的话我感觉，还是要提前抢的话可以解决，我bw都是开票后抢，都是429，抢bml提前抢就没有这提示，也可能是bml抢到人原本就少

**Reply by Colin-Chen630** [reply]
我两次抢票验证码均在5s过完，开始发送请求，bw大会员那一波一直429，但是7.1的联票一次拿下。我觉得这个还是跟服务器压力有关，抢bw的人多，联票人少。全靠运气挤进去。提前抢触发风控得不偿失了

**Reply by Colin-Chen630** [reply]
还有个朋友的抢票经历。他也是一开脚本+手抢。他电脑连接家里网络，手机用的流量，最后手抢抢到，没有出现验证码。我寻思通常流量和家里路由不会相同ip，但是他手机端也没有验证。所以是否真是通过ip白名单，我也不确定

**Reply by SouthSeafireworks** [reply]
确实是这样，但是不太理解为啥会一直429

**Reply by Cloud1069** [reply]
根据我以前的经验429是延迟太低的关系，
也有可能白名单是根据账号来的。

**Comment by gpc123456** [comment]
我这边又测试了几次，这个方法在bw二次放票的时候有可能不可行，因为到时候会刷新票仓，很可能并不会有已售罄的票，而全部都为还未开票的票。此外，bw和bml票仓的白名单并不通用，无法使用其中一个票仓帮助另一个票仓通过验证。因此这个功能应该会作为一个可选辅助功能，如果到时候能碰上合适的场景可以选择使用，不会加入到通用的抢票流程中。

**Reply by mikumifa** [reply]
bw和bw+bml联票可以吗

**Reply by Colin-Chen630** [reply]
我也测试了一下。现在只有在购买联票时会触发验证后显示售空，但是如果选择发送bw普票请求，会在不触发验证的情况下显示活动收摊。我觉得这个方法能不能用取决于到时候向联票发送请求会不会返回活动收摊，如果是这样，那可能就不行了。

**Reply by mikumifa** [reply]
> 我也测试了一下。现在只有在购买联票时会触发验证后显示售空，但是如果选择发送bw普票请求，会在不触发验证的情况下显示活动收摊。我觉得这个方法能不能用取决于到时候向联票发送请求会不会返回活动收摊，如果是这样，那可能就不行了。

二次放票试一下吧

**Reply by Colin-Chen630** [reply]
到时候开票前，找个小号尝试一下

**Reply by Cloud1069** [reply]
我认为要看这个票种的状态吧，目前联票处于官宣不补只会有回流的状态才会报验证码
普票应该处于不补回流准备二开的状态，锁了请求才会有活动收摊的状态吧

**Reply by gpc123456** [reply]
对的对的，这个方法的关键是票仓中要有已售罄的票种，而不是只有不可售的票种，像是昨天抢联票的场景就比较适合

**Reply by Colin-Chen630** [reply]
> 我认为要看这个票种的状态吧，目前联票处于官宣不补只会有回流的状态才会报验证码 普票应该处于不补回流准备二开的状态，锁了请求才会有活动收摊的状态吧

如果是这个逻辑，是个好消息。他如果是退票会回流，那联票可能一直是这个可以请求验证码的状态，那就可以了

**Reply by gpc123456** [reply]
如果官方锁屏重置票仓这个方法就用不了了。我看有的脚本是提前1小时向未开票票仓发请求提前过验证的，不过感觉风险比较大

**Reply by Colin-Chen630** [reply]
> 如果官方锁屏重置票仓这个方法就用不了了。我看有的脚本是提前1小时向未开票票仓发请求提前过验证的，不过感觉风险比较大

但我觉得这个也就开票前手动测才能说得清改没改，如果一律自动走这个方案，触发412就寄喽（

**Reply by mikumifa** [reply]
联票是关键了，联票官方说联票只有一次，第二次抢票联票肯定是在的

**Reply by Cloud1069** [reply]
我没在发售前跑过程序，不知道状态，我都是发售后开跑的，bw二开必然要锁屏重置，就是不知道到时候请求联票是否有效，提前一小时也没用白名单应该坚持不了一小时

**Reply by mikumifa** [reply]
> bw二开必然要锁屏重置

锁屏重置是啥

**Reply by Colin-Chen630** [reply]
> 我没在发售前跑过程序，不知道状态，我都是发售后开跑的，bw二开必然要锁屏重置，就是不知道到时候请求联票是否有效，提前一小时也没用白名单应该坚持不了一小时

可以提前一小时或是一些时间找个小号请求一下看看能不能触发验证，可以的话到时候提前三分钟请求一次把验证过了，然后正常抢

**Reply by gpc123456** [reply]
> 联票是关键了，联票官方说联票只有一次，第二次抢票联票肯定是在的

有道理的，既然官方没说联票会二开，应该不会把联票撤下去

**Reply by Cloud1069** [reply]
> > bw二开必然要锁屏重置
> 
> 锁屏重置是啥

二开上了之后在下面不就有显示时间了么，现在的状态是已售罄

**Reply by Cloud1069** [reply]
关键在于二开时间公布后，改了库存和页面，再次请求联票是否有效了，如果还跟现在的状态一样，这个方法就可以试试了

**Reply by gpc123456** [reply]
> 关键在于二开时间公布后，改了库存和页面，再次请求联票是否有效了，如果还跟现在的状态一样，这个方法就可以试试了

对的。而且通过api拿到的数据比官方前端展示的多，我记得昨天开联票的时候普票好像还是已售罄，只要官方没封这个票种的票仓就行

**Reply by CinnabarCHU** [reply]
我在想，收摊的SKU能不能触发？比如说我拿大会员限定的那个SKU去请求，会发生什么？不知道大大测试过吗
Edit:
没看到折叠的评论，大大测试过了，但是今天我发现至少BW的12日VIP票请求过去还是触发验证码显示售空了，这是运维疏忽了没更新状态？

**Reply by gpc123456** [reply]
没有呢，我就一个账号不太敢试，也没留当时大会员限定的config文件，讨论里有小号和配置文件的朋友可以试一下。

**Reply by Colin-Chen630** [reply]
> 我在想，收摊的SKU能不能触发？比如说我拿大会员限定的那个SKU去请求，会发生什么？不知道大大测试过吗 Edit: 没看到折叠的评论，大大测试过了，但是今天我发现至少BW的12日VIP票请求过去还是触发验证码显示售空了，这是运维疏忽了没更新状态？

![image](https://github.com/mikumifa/biliTickerBuy/assets/87928537/97d1f5ed-8731-43a7-a03c-0a5acc025fad)
我去查了log，两个sku id好像一致。

然后你说的今天发请求显示售空是今天的事，昨天我那么发显示活动收摊了

**Reply by gpc123456** [reply]
确实哎，感觉B站好像直接就是把大会员的票仓直接继承下来了，修改了下起止时间，并没有开新的票仓 ( •̀ ω •́ )✧

**Reply by CinnabarCHU** [reply]
> > 我在想，收摊的SKU能不能触发？比如说我拿大会员限定的那个SKU去请求，会发生什么？不知道大大测试过吗 Edit: 没看到折叠的评论，大大测试过了，但是今天我发现至少BW的12日VIP票请求过去还是触发验证码显示售空了，这是运维疏忽了没更新状态？
> 
> ![image](https://private-user-images.githubusercontent.com/87928537/345368198-97d1f5ed-8731-43a7-a03c-0a5acc025fad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAwMDA5NDksIm5iZiI6MTcyMDAwMDY0OSwicGF0aCI6Ii84NzkyODUzNy8zNDUzNjgxOTgtOTdkMWY1ZWQtODczMS00M2E3LWEwM2MtMGE1YWNjMDI1ZmFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzAzVDA5NTcyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBiYTQ2MDgzNzNmMTM1ZGVmMGJjOWY3Y2U0M2M0MWMzOTg2MmVhNmNmZjNjZTRjMjg5ODkyZDFjYTY3MGY5YzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.LEUgYtaS-vYaKBholP0GgbgiCiGPzQ0h-gqyuavOdFo) 我去查了log，两个sku id好像一致。
> 
> 然后你说的今天发请求显示售空是今天的事，昨天我那么发显示活动收摊了

了解了，那就只能看二开的时候运维怎么操作了……

**Comment by gpc123456** [comment]
与主题无关的信息大家就不要回了呦，例如”天才“，”🐂🍺“等。您发送的消息会通过邮件发送给所有参与讨论的成员，会对大家造成打扰哒，感谢您的理解~

**Comment by JYK89** [comment]
这个就要看提前抢的链接了

**Comment by CinnabarCHU** [comment]
看到已经merge了，我提醒一下，一定谨慎选取提前过码的票种，账号异常在一些情况下会导致的HTTP 429始终无法解除……我到现在也没搞清楚这个规律到底是什么，有的时候账号异常可以无视，有的时候就不行。

**Reply by gpc123456** [reply]
是的，现在对B站账号异常判断的原理还没有非常清楚，建议大家先用小号尝试提前过验证码，只有主力账号时可以先关闭这个功能，以免被风控。我这边测试下来选择已开票/已售罄的票暂时时不会触发风控的，但不清楚其他账号的的情况

**Reply by CinnabarCHU** [reply]
我这边结论一致，未开售的票是最危险的，提前抢票真的是玄学，有的账号提示账号异常但是一过开售时间还是可以正常提交请求，但是有的账号就直接一路被请求被拒绝到死了……

**Reply by Colin-Chen630** [reply]
好像被清算的账号报的是412，一般的拒绝是429

**Reply by MUSESTAR-Official** [reply]
我的是抢联票时抢到一张以后报了412

**Reply by CinnabarCHU** [reply]
> 好像被清算的账号报的是412，一般的拒绝是429

是的，我检查了一下log确实是这样，最好别赌，412的冷却好像是10 min。
